### ISSUE TITLE
[Feature Request] Tapping into events for scripting

### ISSUE THREAD

**Issue by resolritter** [tachiyomiorg/tachiyomi#3865]
## Why/User Benefit/User Problem

Disclaimer: this would not incur any extra frontend change as I'm proposing it. It's more of a low-level backend change aimed at "power users".

The motivating use-case is syncing. Tachiyomi has automatic tracking for some providers (e.g. AniList), but I would like to have my own setup for dealing with events such as "Series completed" or "Nth chapter has been read".

Android has automation apps such as Tasker which can listen to intents and file changes. File changes wouldn't be useful currently because everything is written to the database, but I was surprised to not find anything useful in the logs either.

![Screenshot_2020-09-29-22-53-48-323_com pluscubed matloglibre_1](https://user-images.githubusercontent.com/17429390/94634711-26485280-02a7-11eb-9507-f39c6419d792.png)

What about database triggers? Could I, as a user, somehow listen to some table's changes and go from there? According to this [SO answer](https://stackoverflow.com/questions/24190615/how-to-listen-for-sqlite-database-changes-in-android), SQLite on Android allegedly does not support it. It's also sort of a "backdoor" approach because it would be injecting SQL code for adding a behavior to the database.

After all my considerations, the best thing available currently would be to "poll" the database tables every N minutes using e.g. Tasker. This still strictly worse than tapping into the events as actions are performed in the backend.

## What/Requirements

1. Implement a logging abstraction for events. Make it so it's straightforward to support different kinds of events in the future.
2. Initially, log the following operations. Make them easy to parse for scripts.
- Chapter was read
- Series was completed
3. Optional: Gate this logging behind a feature flag. Could be a UI checkbox. The reason why I'm classifying it "optional" is that it isn't costly to log those events.

Finally, it's possible that I haven't considered a way to achieve this currently. If so, I'd like to know how. Thanks in advance.

**Comment by arkon** [comment]
> but I was surprised to not find anything useful in the logs either.

You should be more surprised by the amount of junk that many apps are outputting into the log stream. It's one thing if an app is excessively logging things for tracing reasons, but we don't upload Tachiyomi logs other than crash reports, hence the general lack of logging.

**Comment by resolritter** [comment]
> > but I was surprised to not find anything useful in the logs either.
> 
> You should be more surprised by the amount of junk that many apps are outputting into the log stream. It's one thing if an app is excessively logging things for tracing reasons, but we don't upload Tachiyomi logs other than crash reports, hence the general lack of logging.

@arkon  Fair enough, thanks for the answer. I suggested that logging should be gated by a feature flag, as mentioned, but that would add some complexity and incur frontend changes.

About the feature: I own both Automate and Tasker and can confirm both can easily tap into the logs with some matching rules. As mentioned, both apps also support listening to file changes, thus appending to a log file in the sdcard also would work... In fact, it would be even easier to interact with it that way; couple of upsides

1. No need create expressions for filtering
2. You can process the event history at any time, from any point, thus nothing is ever missed

However, that does sound more complex than just printing some logs. If you have an opinion between one or the other, please let me know.

**Comment by resolritter** [comment]
Don't know Kotlin but I tried to take a look at this today.

# "Chapter was read" use-case

https://github.com/inorichi/tachiyomi/blob/ce0090f0cab2619703ff903ad80a188bc62e5214/app/src/main/java/eu/kanade/tachiyomi/data/database/queries/ChapterQueries.kt#L87

https://github.com/inorichi/tachiyomi/blob/ce0090f0cab2619703ff903ad80a188bc62e5214/app/src/main/java/eu/kanade/tachiyomi/data/database/queries/ChapterQueries.kt#L92

And, from there

https://github.com/inorichi/tachiyomi/blob/ce0090f0cab2619703ff903ad80a188bc62e5214/app/src/main/java/eu/kanade/tachiyomi/data/database/resolvers/ChapterProgressPutResolver.kt#L19

# "Series was completed" use-case

This one, I'm guessing, can be figured out by querying the database directly and seeing if `readCount` === `chapterCount` whenever `"Chapter was read"` gets noticed, thus it might not necessarily need an implementation.

**Comment by resolritter** [comment]
I changed my mind about this a long time ago, but forgot to update here.

I wanted to listen to log events so that I could use them for my own tracking system, however I later realized that creating a custom API which matches one of the supported ones (e.g. AniList) and integrating it into Tachiyomi would probably be better. Even if that's also not supported at the moment, it probably is a saner model to follow and likely would require less changes.

I don't have a use-case for the original suggestion anymore. Closing.
